src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,188 @@
+
+import static java.time.Instant.now;
+
+/**
+ * SamRetry encapsulates logic needed for retrying Sam API calls. The retry function is linear and
+ * all constants are hard-coded, as Sam failures are generally just short periods of down-time.
+ */
+public class SamRetry {
+    private static Logger logger = LoggerFactory.getLogger(SamRetry.class);
+    private static final int MAXIMUM_WAIT_SECONDS = 30;
+    private static final int INITIAL_WAIT_SECONDS = 10;
+    private static final int OPERATION_TIMEOUT_SECONDS = 300;
+    private final Instant operationTimeout;
+    private int retrySeconds;
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    SamRetry() {
+        this.operationTimeout = now().plusSeconds(OPERATION_TIMEOUT_SECONDS);
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @VisibleForTesting
+    SamRetry(int timeoutSeconds) {
+        this.operationTimeout = now().plusSeconds(timeoutSeconds);
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @FunctionalInterface
+    public interface SamVoidFunction {
+        void apply() throws ApiException, InterruptedException;
+    }
+
+    @FunctionalInterface
+    public interface SamFunction<R> {
+        R apply() throws ApiException, InterruptedException;
+    }
+
+
+    static <T> T retry(SamFunction<T> function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        return samRetry.perform(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static <T> T retry(SamFunction<T> function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        return samRetry.perform(function);
+    }
+
+    static void retry(SamVoidFunction function)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry();
+        samRetry.performVoid(function);
+    }
+
+    // Version of the retry function for testing so that we can override timeout seconds
+    @VisibleForTesting
+    static void retry(SamVoidFunction function, int timeoutSeconds)
+        throws InterruptedException {
+        SamRetry samRetry = new SamRetry(timeoutSeconds);
+        samRetry.performVoid(function);
+    }
+
+    private <T> T perform(SamFunction<T> function) throws InterruptedException {
+        while (true) {
+            try {
+                return function.apply();
+            } catch (ApiException ex) {
+                handleApiException(ex);
+            } catch (Exception ex) {
+                throw new InternalServerErrorException("Unexpected exception type: " + ex.toString(), ex);
+            }
+            sleepBeforeRetrying();
+        }
+    }
+
+    private void performVoid(SamVoidFunction function) throws InterruptedException {
+        while (true) {
+            try {
+                function.apply();
```

Comment:
consider wrapping SamVoidFunction as a `SamFunction<Void>` and calling `perform` instead to cut down on duplicated code. e.g.
```
private void performVoid(SamVoidFunction function) throws InterruptedException {
  perform(() -> { function.apply(); return null; }); 
}

Reply:
Nice!